File: medplum-java/medplum-java-model/src/main/java/com/fhir/Graphdefinition_compartmentRule.java
package com.fhir;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Locale;
import javax.annotation.Nonnull;

public final class Graphdefinition_compartmentRule {
  public static final Graphdefinition_compartmentRule MATCHING = new Graphdefinition_compartmentRule(Value.MATCHING, "MATCHING");

  public static final Graphdefinition_compartmentRule CUSTOM = new Graphdefinition_compartmentRule(Value.CUSTOM, "CUSTOM");

  public static final Graphdefinition_compartmentRule DIFFERENT = new Graphdefinition_compartmentRule(Value.DIFFERENT, "DIFFERENT");

  public static final Graphdefinition_compartmentRule IDENTICAL = new Graphdefinition_compartmentRule(Value.IDENTICAL, "IDENTICAL");

  private final Value value;

  private final String string;

  Graphdefinition_compartmentRule(Value value, String string) {
    this.value = value;
    this.string = string;
  }

  public Value getEnumValue() {
    return value;
  }

  @Override
  @JsonValue
  public String toString() {
    return this.string;
  }

  @Override
  public boolean equals(Object other) {
    return (this == other) 
      || (other instanceof Graphdefinition_compartmentRule && this.string.equals(((Graphdefinition_compartmentRule) other).string));
  }

  @Override
  public int hashCode() {
    return this.string.hashCode();
  }

  public <T> T accept(Visitor<T> visitor) {
    switch (value) {
      case MATCHING:
        return visitor.visitMATCHING();
      case CUSTOM:
        return visitor.visitCUSTOM();
      case DIFFERENT:
        return visitor.visitDIFFERENT();
      case IDENTICAL:
        return visitor.visitIDENTICAL();
      case UNKNOWN:
      default:
        return visitor.visitUnknown(string);
    }
  }

  @JsonCreator(
      mode = JsonCreator.Mode.DELEGATING
  )
  public static Graphdefinition_compartmentRule valueOf(@Nonnull String value) {
    String upperCasedValue = value.toUpperCase(Locale.ROOT);
    switch (upperCasedValue) {
      case "MATCHING":
        return MATCHING;
      case "CUSTOM":
        return CUSTOM;
      case "DIFFERENT":
        return DIFFERENT;
      case "IDENTICAL":
        return IDENTICAL;
      default:
        return new Graphdefinition_compartmentRule(Value.UNKNOWN, upperCasedValue);
    }
  }

  public enum Value {
    IDENTICAL,

    MATCHING,

    DIFFERENT,

    CUSTOM,

    UNKNOWN
  }

  public interface Visitor<T> {
    T visitIDENTICAL();

    T visitMATCHING();

    T visitDIFFERENT();

    T visitCUSTOM();

    T visitUnknown(String unknownType);
  }
}

File: medplum-typescript/src/model/fhir/Graphdefinition_compartmentRule.ts
export type Graphdefinition_compartmentRule =
    | Graphdefinition_compartmentRule.Identical
    | Graphdefinition_compartmentRule.Matching
    | Graphdefinition_compartmentRule.Different
    | Graphdefinition_compartmentRule.Custom;

export const Graphdefinition_compartmentRule = {
    Identical: "identical" as Graphdefinition_compartmentRule.Identical,
    Matching: "matching" as Graphdefinition_compartmentRule.Matching,
    Different: "different" as Graphdefinition_compartmentRule.Different,
    Custom: "custom" as Graphdefinition_compartmentRule.Custom,

    _visit: <Result>(value: Graphdefinition_compartmentRule, visitor: Graphdefinition_compartmentRule._Visitor<Result>): Result => {
        switch (value) {
            case Graphdefinition_compartmentRule.Identical: return visitor.identical();
            case Graphdefinition_compartmentRule.Matching: return visitor.matching();
            case Graphdefinition_compartmentRule.Different: return visitor.different();
            case Graphdefinition_compartmentRule.Custom: return visitor.custom();
            default: return visitor._unknown();
        }
    },
};

export declare namespace Graphdefinition_compartmentRule {
    export type Identical = "identical" & {
        __Identical: void,
    };
    export type Matching = "matching" & {
        __Matching: void,
    };
    export type Different = "different" & {
        __Different: void,
    };
    export type Custom = "custom" & {
        __Custom: void,
    };

    export interface _Visitor<Result> {
        identical: () => Result;
        matching: () => Result;
        different: () => Result;
        custom: () => Result;
        _unknown: () => Result;
    }
}
